### PR TITLE
refactor: switch to uv for dependency management and build

### LINKED ISSUE #238: Switch to uv for dependency management and build
Currently the repo uses `setuptools` to build the pipeline and needs to be updated to use [uv](https://docs.astral.sh/uv/). Switching to uv will simplify build, dependency management, and the developer contributions, as well as reduce the number of config files in the top directory. This change may also allow the package to be used with multiple Python version, since uv's dependency solver can account for this. 

This will require:
- Picking a particular version of uv to use[^1]
- Setting up uv so that the `pyproject.toml` file contains package information and all dependencies
- Adding the resulting `uv.lock` file that includes the solved dependencies
- Removing any other files used for the old dependencies, e.g. `setup.cfg`, `setup.py`
- Moving any configuration information that can be into `pyproject.toml`, e.g. ruff rules/ignores
- Editing the Github Actions workflow file `deploy_pypi.yaml` to use uv to build the package and deploy to PyPi
- Remove bumpversion configs/etc. since versioning should be handled centrally by uv, and the `uv version` command

[^1]: The latest available version as of writing is 0.11.7, whereas I've been using 0.7.19 since that was the latest version when I first started using it.

### PR DESCRIPTION
**Pull request by humnabaa**
## Description
<details><summary>Please include a summary of the changes. </summary>

  - Switches the project from setuptools/pip to uv for dependency management, builds, and CI.
  - Consolidates 6 config files (setup.cfg, setup.py, .bumpversion.cfg, .ruff.toml, .isort.cfg, old pyproject.toml) into a single pyproject.toml.
  - Updates all three GitHub Actions workflows to use uv.
  - Replaces black + isort pre-commit hooks with ruff-format.
  - Updates contribution_guide.md and branch_and_deploy_guide.md.
  - This is a breaking change for the developer workflow (uv replaces pip) but does NOT break any end-user functionality.
  - Tested locally: all non-Spark tests pass, ruff reports no issues.
</details>

Closes #238

## Type of change

- [x] Breaking change - *backwards incompatible change, changes expected behaviour*
- [x] Non-user facing change, structural change, dev functionality, docs ...

## Checklist:

- [x] I have performed a self-review of my own code.
- [x] I have commented my code appropriately, focusing on explaining my design decisions (explain why, not how).
- [x] I have made corresponding changes to the documentation (comments, docstring, etc.. )
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] I have updated the change log.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -1,3 +1,135 @@
...
+    "PD",
+    "PT",
+    "PTH",
+    "Q",
+    "T10",
+    "T20",
+    "W",
+    "YTT",
+]
+ignore = ["N812", "E712", "E721", "COM812"]
```
Comment (nclarkz) [inline pyproject.toml]: Missing "PD901" which is failing on the github precommit checks.

File: rdsa_utils/helpers/pyspark_log_parser/parser.py [hunk 2]
```diff
@@ -167,9 +167,12 @@ def parse_pyspark_logs(
...
             props = event.get("Properties", {})
-            mem, overhead = props.get("spark.executor.memory", "0g"), props.get(
-                "spark.yarn.executor.memoryOverhead",
-                "0g",
+            mem, overhead = (
+                props.get("spark.executor.memory", "0g"),
+                props.get(
+                    "spark.yarn.executor.memoryOverhead",
+                    "0g",
+                ),
```
Comment (coatet) [inline rdsa_utils/helpers/pyspark_log_parser/parser.py]: Was this a ruff suggestion? Not seen this way of defining a tuple before?

File: .github/workflows/pull_request_workflow.yaml
Comment (coatet) [inline .github/workflows/pull_request_workflow.yaml]: This step should use uv now, probably `uv sync --group dev`

File: .github/workflows/deploy_pypi.yaml [hunk 3]
```diff
@@ -10,21 +10,17 @@ jobs:
...
-      run: |
-        python -m pip install --upgrade pip
-        pip install setuptools wheel build 
-
-    - name: Build the package
-      run: |
-        python -m build
-
+    - name: Build package
+      run: uv build
```
Comment (coatet) [inline .github/workflows/deploy_pypi.yaml]: Have you checked the format we need for PyPi? In the past I've used `--wheel` but that was for Artifactory, so we should check this.

File: .github/workflows/deploy_pypi.yaml
Comment (coatet) [inline .github/workflows/deploy_pypi.yaml]: I haven't tried it yet but uv now has a [uv publish](https://docs.astral.sh/uv/guides/package/#publishing-your-package) command that should do this, might be cleaner?

File: .github/workflows/pull_request_workflow.yaml
Comment (coatet) [inline .github/workflows/pull_request_workflow.yaml]: This may need to switch to something like `uv run pytest`?

Comment (nclarkz) [inline .github/workflows/pull_request_workflow.yaml]: Needs to be changed to uv and check pytest then works

**Review by nclarkz — CHANGES_REQUESTED** [review]
Couple of updates due to failing github checks

File: rdsa_utils/test_utils.py [hunk 4]
```diff
@@ -134,7 +134,7 @@ def test(foo, bar):
         args = case.kwargs.keys()
 
         # Make sure all keys are in each case, otherwise initialise with None.
-        diff = {k: None for k in set(all_args) - set(args)}
+        diff = dict.fromkeys(set(all_args) - set(args))
```
Comment (coatet) [inline rdsa_utils/test_utils.py]: Again, was this a ruff suggestion? Would prefer to keep any actual code changes minimal for this PR.

**Comment by nclarkz** [comment]
##  Code review

#### Documentation

Any new code includes all the following forms of documentation:

- [x] **Function Documentation** as docstrings within the function definition.
- [x] **Examples** demonstrating major functionality, which runs successfully locally.
- [x] **Changelog** Ticket added to changelog in correct place.

#### Functionality

- [ ] **Pipeline success**: Pipeline can be successfully run for all expected instances of the pipeline.
- [ ] **Functionality**: Any functional claims of new code have been confirmed.
- [x] **Parameters**: Parameters adhere to a parsimonious principle, without redundancy or unnecessary dynamism.
- [x] **Automated tests**: Unit tests cover essential functions for a reasonable range
  of inputs and edge case conditions. All tests pass on your local machine.
- [x] **Packaging guidelines**: New code conforms to the project contribution
  guidelines.

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

Currently not passing github actions checks, that needs to be rectified prior to merging. Work is forked rather than branched, so need to clone the separate repo to test it works my end once it is passing the checks, would be worth branching rather than forking in the future for ease of peer review.

Will need to update some of the documentation to reflect differing instructions for uv:
- docs/branch_and_deploy_guide.md needs to change bump2version to uv version instructions
- docs/contribution_guide.md update virtual environment section, setup dev environment change to pip install uv, uv init, installation of pre-commit needs to be checked whether the uv install captures it all
- Don't think the readme needs updating as the installation instructions are for the package, so it's ran at the user's end?
- Does a uv.lock file need to be added?

File: .pre-commit-config.yaml [hunk 5]
```diff
@@ -18,23 +18,12 @@ repos:
...
-  - repo: https://github.com/psf/black
-    rev: 24.10.0
-    hooks:
-      - id: black
-
-  - repo: https://github.com/PyCQA/isort
-    rev: 5.13.2
-    hooks:
-      - id: isort
-        args: ["--profile", "black"]
```
Comment (coatet) [inline .pre-commit-config.yaml:31]: Looks like you've removed isort, which we definitely want to keep.

**Review by coatet — CHANGES_REQUESTED** [review]

**Comment by coatet** [comment]
Something that's occurred to me because I've been looking at pandas versions for something else: your `uv.lock` has pegged a fairly old version of pandas, and I'm concerned that this change will force anyone who installs rdsa-utils to use pandas 2.0. Version 3.0 is available and has some useful updates, so we need to test what uv does when you list rdsa-utils as a dependency. The pyproject file doesn't specify a pandas version, so I'd like to know whether uv forces you onto 2.0 or lets the pandas version vary up to 3.0+. 

For reference, the big difference in [pandas 3.0](https://pandas.pydata.org/docs/dev/whatsnew/v3.0.0.html) is that it uses a `string` dtype instead of `Object` for non-numerical columns by default (which was causing validation issues on my current pipeline) and you can use `pd.col()` to create new columns in a more Pyspark-like way rather than using lambda functions. I think these are important changes so I don't want to force anyone to use pandas 2.0 just to use rdsa-utils.

File: pyproject.toml (hunk 1 above)
Comment (humnabaa) [reply pyproject.toml]: Added "PD901" to the ruff ignore list in [pyproject.toml]

File: .github/workflows/pull_request_workflow.yaml
Comment (humnabaa) [reply .github/workflows/pull_request_workflow.yaml]: now uses uv sync --group dev for installation and uv run pytest for running tests.

File: rdsa_utils/helpers/pyspark_log_parser/parser.py (hunk 2 above)
Comment (humnabaa) [reply rdsa_utils/helpers/pyspark_log_parser/parser.py]: Yes, it was a ruff auto-format suggestion. I've reverted this back to the original style so that it keeps code changes minimal as you suggested.

File: .github/workflows/pull_request_workflow.yaml
Comment (humnabaa) [reply .github/workflows/pull_request_workflow.yaml]: switched this to uv sync --group dev + [uv run pytest]

File: .github/workflows/deploy_pypi.yaml (hunk 3 above)
Comment (humnabaa) [reply .github/workflows/deploy_pypi.yaml]: uv build produces both sdist (.tar.gz) and wheel (.whl) by default which is the standard expected by PyPI.

File: .github/workflows/deploy_pypi.yaml
Comment (humnabaa) [reply .github/workflows/deploy_pypi.yaml]: Good call . switched to uv publish with UV_PUBLISH_TOKEN env var

File: .github/workflows/pull_request_workflow.yaml
Comment (humnabaa) [reply .github/workflows/pull_request_workflow.yaml]: Done

File: .pre-commit-config.yaml (hunk 5 above)
Comment (humnabaa) [reply .pre-commit-config.yaml:31]: Added it back in

**Comment by humnabaa** [comment]
@nclarkz Updated [branch_and_deploy_guide.md](vscode-file://vscode-app/c:/Program%20Files/Microsoft%20VS%20Code/8761a5560c/resources/app/out/vs/code/electron-browser/workbench/workbench.html) (bump2version is updated to manual version update in __init__.py] and contribution_guide.md (done a full rewrite of setup section to use uv sync --group dev, uv run pytest, uv run pre-commit install)

**Comment by humnabaa** [comment]
@coatet checked the uv.lock and it resolves correctly per Python version: pandas 2.0.3 only for Python <3.9, pandas 2.3.3 for 3.9–3.10, and pandas 3.0.2 for Python ≥3.11. Since the pyproject.toml has no version pin on pandas, uv.lock won't constrain downstream users. they willl resolve their own pandas version when they pip install rdsa-utils. The lock file only applies to development.

File: rdsa_utils/helpers/pyspark_log_parser/parser.py [hunk 6]
```diff
@@ -167,12 +167,9 @@ def parse_pyspark_logs(
...
             props = event.get("Properties", {})
-            mem, overhead = (
-                props.get("spark.executor.memory", "0g"),
-                props.get(
-                    "spark.yarn.executor.memoryOverhead",
-                    "0g",
-                ),
+            mem, overhead = props.get("spark.executor.memory", "0g"), props.get(
+                "spark.yarn.executor.memoryOverhead",
+                "0g",
```
Comment (coatet) [inline rdsa_utils/helpers/pyspark_log_parser/parser.py]: Can we revert this back to how it was? This is failing a ruff-format pre-commit check and there was no need to change.

File: rdsa_utils/test_utils.py [hunk 7]
```diff
@@ -134,7 +134,7 @@ def test(foo, bar):
         args = case.kwargs.keys()
 
         # Make sure all keys are in each case, otherwise initialise with None.
-        diff = dict.fromkeys(set(all_args) - set(args))
+        diff = {k: None for k in set(all_args) - set(args)}
```
Comment (coatet) [inline rdsa_utils/test_utils.py]: Why was this changed? Was this an automatic ruff thing? There may be a good reason for this but I would prefer to keep functional changes to a minimum in this PR.

**Review by coatet — COMMENTED** [review]
@humnabaa Looks all good now! The CI/CD is failing for two reasons which should be quick to fix:
- The `ruff-format` pre-commit check is failing on `helpers/pyspark_log_parser/parser.py` because of some odd reformatting. I've pointed it out below and think it should just be reverted. 
- All the builds are failing because Github Actions can't access our Artifactory, since that's where the uv.lock says to find the packages. This is correct for users but won't work for CI/CD. The solution I've used for this is to tell Github Actions to re-solve the dependencies using PyPi just before the sync. Adding `uv lock --default-index https://pypi.org/simple` just before the `uv sync` step should fix it.

File: rdsa_utils/helpers/pyspark_log_parser/parser.py (hunk 6 above)
Comment (coatet) [reply rdsa_utils/helpers/pyspark_log_parser/parser.py]: Just realised this was changed, sorry! See what ruff gives you when you do `ruff check --fix` and we can probably just accept that.

Comment (humnabaa) [reply rdsa_utils/helpers/pyspark_log_parser/parser.py]: ran ruff check --fix and it reports all checks passed with no further changes

File: rdsa_utils/test_utils.py (hunk 7 above)
Comment (humnabaa) [reply rdsa_utils/test_utils.py]: That change wasn’t intentional for behavior and is functionally equivalent, but I agree it adds unnecessary noise to this PR. Reverted this back

**Review by coatet — APPROVED** [review]
All looking good now!

**Review by nclarkz — APPROVED** [review]
